tests/test_checkout.py

```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
     def test_two_write_checkouts(self, repo):
         w1_checkout = repo.checkout(write=True)
         with pytest.raises(PermissionError):
-            w2_checkout = repo.checkout(write=True)
+            repo.checkout(write=True)
         w1_checkout.close()
 
     def test_two_read_checkouts(self, repo, array5by7):
         w_checkout = repo.checkout(write=True)
         dataset_name = 'dset'
         r_ds = w_checkout.datasets.init_dataset(name=dataset_name, prototype=array5by7)
         r_ds['1'] = array5by7
+        w_checkout.metadata.add('init', 'array5by7 added')
         w_checkout.commit('init')
         r1_checkout = repo.checkout()
         r2_checkout = repo.checkout()
         assert np.allclose(r1_checkout.datasets['dset']['1'], array5by7)
-        assert np.allclose(r1_checkout.datasets['dset']['1'], r2_checkout.datasets['dset']['1'])
+        assert np.allclose(
+            r1_checkout.datasets['dset']['1'], r2_checkout.datasets['dset']['1'])
+        assert r1_checkout.metadata.get('init') == 'array5by7 added'
+        assert r2_checkout.metadata.get('init') == 'array5by7 added'
         r1_checkout.close()
         r2_checkout.close()
         w_checkout.close()
 
     def test_write_with_r_checkout(self, written_repo, array5by7):
         co = written_repo.checkout()
+        # TODO: This should raise more informative error message
+        # Now it's type error because `init_dataset` is None
         with pytest.raises(TypeError):
             co.datasets.init_dataset(name='dset', shape=(5, 7), dtype=np.float64)
-            co.datasets['_dset']['1'] = array5by7
-            co.datasets['_dset'].add('1', array5by7)
+        # TODO: This should be permission error
+        with pytest.raises(AttributeError):
+            co.metadata.add('a', 'b')
         co.close()
 
     def test_write_after_repo_deletion(self, written_repo, array5by7):
         co = written_repo.checkout(write=True)
         shutil.rmtree(written_repo._repo_path)
         with pytest.raises(OSError):
             co.datasets.init_dataset('dset', prototype=array5by7)
+        co.metadata.add('a', 'b')
         dset = co.datasets['_dset']
         dset['1'] = array5by7
-        # todo: shouldn't it be atleast same as above
         with pytest.raises(FileNotFoundError):
             co.commit()
         co.close()
 
-
-@pytest.mark.skip(reason='not implemented')
-class TestBranching(object):
-
-    def test_merge(self):
-        pass
-
-    def test_name_conflict(self):
-        pass
-
-    def test_delete_branch(self):
-        pass
-
-    def test_merge_multiple_checkouts(self):
-        pass
-
-    def test_merge_diverged(self):
-        pass
-
-    def test_merge_diverged_conflict(self):
-        pass
+    def test_operate_on_dataset_after_closing_old_co(self, repo, array5by7):
```

Comment:
style note: don't use abreviations in test case function names. I don't care if test case function names go over some arbitrary line length limit (90ish for the main code base) since they should be as descriptive as possible.